### PR TITLE
Plumbing functions for walking working directory

### PR DESCRIPTION
**Pull request by seanses**
Implements two functions: `walk_working_dir` for walking a repo working directory; `filter_files_from_index` for filtering out untracked files. Needed for `materialize` and `dematerialize` commands.

### CODE REVIEW

**Comment by ylow** [comment]
If helpful we have used the walkdir package elsewhere https://docs.rs/walkdir/latest/walkdir/

**Comment by seanses** [comment]
> If helpful we have used the walkdir package elsewhere https://docs.rs/walkdir/latest/walkdir/

Sure. Simple change.

File: rust/gitxetcore/src/git_integration/git_repo_plumbing.rs [hunk 1]
```diff
@@ -357,14 +359,148 @@ pub fn list_files_from_repo(
...
+
+        let dir_walker = std::fs::read_dir(sr)?;
+
+        for entry in dir_walker.flatten() {
+            let file_type = entry.file_type()?;
+            let file_path = entry.path();
+
+            if file_type.is_file() {
+                ret.push(file_path.strip_prefix(&work_root)?.to_owned());
+            } else if recursive
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/git_repo_plumbing.rs]: I think we need a check here for is_dir(); if it's a symlink, it ends up getting put on the queue and treated as a directory.

Comment (seanses) [reply rust/gitxetcore/src/git_integration/git_repo_plumbing.rs]: Good point. I've rewritten this part using walkdir, should exclude symlinks.

File: rust/gitxetcore/src/git_integration/git_repo_plumbing.rs [hunk 2]
```diff
@@ -357,14 +359,135 @@ pub fn list_files_from_repo(
...
+        return Ok(vec![search_root.strip_prefix(work_root)?.to_owned()]);
+    }
+
+    // now search_root is a regular directory, walk below it
+
+    let ret: Vec<_> = WalkDir::new(&search_root)
+        .follow_links(false)
+        .max_depth(if recursive { usize::MAX } else { 1 })
+        .into_iter()
+        .filter_entry(|e| !is_git_special_files(e.file_name().to_str().unwrap_or_default()))
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/git_repo_plumbing.rs:413]: Does this actually exclude the files under .git/ with the walkdir iter now?  It has to be an exact match, so I think the is_git_special_files function needs to test the prefix.

Comment (hoytak) [reply rust/gitxetcore/src/git_integration/git_repo_plumbing.rs:413]: Ah, I see that it does.  Nice.

File: rust/gitxetcore/src/git_integration/git_repo_plumbing.rs [hunk 3]
```diff
@@ -357,14 +359,135 @@ pub fn list_files_from_repo(
...
+            assert_eq!(bytes2path(&_entry.path), f);
+            ret.push(f.to_owned());
+        }
+    }
+
+    Ok(ret)
+}
+
+pub fn is_git_special_files(path: &str) -> bool {
+    matches!(path, ".git" | ".gitignore" | ".gitattributes")
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/git_repo_plumbing.rs:457]: Do we also want to add .xet here?

**Review by hoytak — APPROVED** [review]
Looks good to me!  Minor things.

**Pull request merged**